src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,150 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2016 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import re
+import sys
+import string
+from itertools import chain
+
+import hypothesis.strategies as strats
+from hypothesis.searchstrategy import SearchStrategy
+from hypothesis.internal.compat import hunichr, hrange
+
+
+def strategy_concat(strategies):
+    """given a list of strategies yielding strings return a strategy yielding
+    their concatenation."""
+    return strats.tuples(*strategies).map(lambda x: u"".join(x))
+
+
+class RegexStrategy(SearchStrategy):
+    """Strategy for generating strings matching a regular expression.
+
+    Currently does not support \B, positive and negative lookbehind
+    assertions and conditional matching.
+
+    """
+
+    def __init__(self, pattern):
+        parsed = re.sre_parse.parse(pattern)
+        self.cache = {}
+        self.strategies = [self._handle_state(state) for state in parsed]
+
+    def do_draw(self, data):
+        self.cache = {}
+        return u"".join(data.draw(strat) for strat in self.strategies)
+
+    def _categories(self, category):
+        if category == 'category_digit':
+            return string.digits
+        elif category == 'category_not_digit':
+            return string.ascii_letters + string.punctuation
+        elif category == 'category_space':
+            return string.whitespace
+        elif category == 'category_not_space':
+            return string.printable.strip()
+        elif category == 'category_word':
+            return string.ascii_letters + string.digits + '_'
```

Comment:
I'd guess this method is the source of a lot of the slowness, since these character sets are recomputed on each draw. It would be worth caching them, or just generating them on module import.